File: creational/abstract-factory/typescript/concrete-factory.ts
import { Logger, ConfigReader } from './product';
import { LoggingFactory } from './factory';
import {
	ConsoleConfigReader,
	ConsoleLogger,
	FileConfigReader,
	FileLogger,
} from './concrete-product';

export class FileLoggingFactory implements LoggingFactory {
	createLogger(): Logger {
		return new FileLogger();
	}

	createConfigReader(): ConfigReader {
		return new FileConfigReader();
	}
}

export class ConsoleLoggingFactory implements LoggingFactory {
	createLogger(): Logger {
		return new ConsoleLogger();
	}

	createConfigReader(): ConfigReader {
		return new ConsoleConfigReader();
	}
}

File: creational/abstract-factory/python/concrete_factory.py
from product import Logger, ConfigReader
from factory import LoggingFactory
from concrete_product import (
    FileLogger,
    ConsoleLogger,
    FileConfigReader,
    ConsoleConfigReader,
)


class FileLoggingFactory(LoggingFactory):
    def create_logger(self) -> Logger:
        return FileLogger()

    def create_config_reader(self) -> ConfigReader:
        return FileConfigReader()


class ConsoleLoggingFactory(LoggingFactory):
    def create_logger(self) -> Logger:
        return ConsoleLogger()

    def create_config_reader(self) -> ConfigReader:
        return ConsoleConfigReader()
